lang/csharp/src/apache/main/AvroDecimal.cs

```diff
@@ -0,0 +1,929 @@
+        /// </summary>
+        /// <param name="value">The unsigned int value.</param>
+        public AvroDecimal(uint value)
+            : this(new BigInteger(value), 0) { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AvroDecimal"/> class from a given unsigned long.
+        /// </summary>
+        /// <param name="value">The unsigned long value.</param>
+        public AvroDecimal(ulong value)
+            : this(new BigInteger(value), 0) { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AvroDecimal"/> class from a given <see cref="BigInteger"/>
+        /// and a scale.
+        /// </summary>
+        /// <param name="unscaledValue">The double value.</param>
+        /// <param name="scale">The scale.</param>
+        public AvroDecimal(BigInteger unscaledValue, int scale)
+        {
+            UnscaledValue = unscaledValue;
+            Scale = scale;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AvroDecimal"/> class from a given array of bytes.
+        /// </summary>
+        /// <param name="value">The byte array.</param>
+        public AvroDecimal(byte[] value)
+        {
+            byte[] number = new byte[value.Length - 4];
+            byte[] flags = new byte[4];
+
+            Array.Copy(value, 0, number, 0, number.Length);
+            Array.Copy(value, value.Length - 4, flags, 0, 4);
+
+            UnscaledValue = new BigInteger(number);
+            Scale = BitConverter.ToInt32(flags, 0);
+        }
+
+        /// <summary>
+        /// Gets the unscaled integer value represented by the current <see cref="AvroDecimal"/>.
+        /// </summary>
+        public BigInteger UnscaledValue { get; }
+
+        /// <summary>
+        /// Gets the sign of the current <see cref="AvroDecimal"/>.
+        /// </summary>
+        public int Sign
+        {
+            get { return UnscaledValue.Sign; }
+        }
+
+        /// <summary>
+        /// Gets the scale of the current <see cref="AvroDecimal"/>.
+        /// </summary>
+        public int Scale { get; }
+
+        /// <summary>
+        /// Converts the current <see cref="AvroDecimal"/> to a string.
+        /// </summary>
+        /// <returns>A string representation of the numeric value.</returns>
+        public override string ToString()
+        {
+            var number = UnscaledValue.ToString(CultureInfo.CurrentCulture);
+
+            if (Scale > 0)
+                return number.Insert(number.Length - Scale, CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator);
+
+            return number;
+        }
+
+        /// <summary>
+        /// Converts the current <see cref="AvroDecimal"/> to a byte array.
+        /// </summary>
+        /// <returns>The current <see cref="AvroDecimal"/> converted to an array of bytes.</returns>
+        public byte[] ToByteArray()
```

Comment:
I think it's better if this wasn't on this class - let the user cast to some other type for serialization - or after all, this is a library exactly for the purpose of serialization! having explicit support for byte[] conversion opens the question of exactly the format, which isn't addressed in the comments, and as noted, there's no need for it.